006-strimzi-ui.md

```diff
@@ -0,0 +1,129 @@
+I would also suggest using Express as a server for this UI. This is due to its small footprint, modularity and available 3rd party modules, such as [helmet for http security](https://helmetjs.github.io/), [Passport.js for authentication and authorization support](http://www.passportjs.org/) or [the graphql and express-graphql modules for GraphQL server support](https://graphql.org/graphql-js/running-an-express-graphql-server/).
+
+This UI would also be provided with a full set of supporting elements - such as end to end tests, automation (examples of which are [in this section](#a-ui-repository-and-ways-of-working-in-the-repository)) to manage common tasks, and documentation around implementation approach, UI best practise and so on.
+
+The below shows how these pieces could integrate.
+
+![Suggested topology](./006-topology.png)
+
+Where:
+
+- 'strimzi-api' is a backend server for Kafka data/requests
+- 'Supporting elements' could be other Kafka related deployments such as Mirror Maker, Kafka Connect and Cruise control. These could also be non Kafka related deployments (not deployed as a part of Strimzi), such as LDAP servers, metric stores etc.
+
+### A UI repository, and ways of working in the repository
+
+In addition to the implementation detailed above, I would propose that any UI contributed to Strimzi be managed and developed as follows:
+
+- Be developed in a dedicated repository inside the Strimzi Github organisation
+- Both the UI server and client are contained in this repo - worked as sub modules in a monorepo
+- That design/discussion/implementation and defect issues are kept solely in this repository
+- All development is done in a behavioural manner - focusing on the end user, and the task they are trying to achieve (see [topic list](#topic-list-page) for an example)
+- That the repository operate in a cloud like CI/CD manner; lots of small/little and often deliveries of new function, appropriately sized and gated, allowing for the UI to be shipped at any time
+- That automation, such as [GitHub actions](https://github.com/features/actions) for example, is used as much as possible to support this CI/CD model - eg, automated dependency updates, automated testing run on PR, issue triage etc
+
+### Phased delivery
+
+A UI is more than just what a user sees. It also needs to be supported by backend data, services, and resources. To this end, and in bearing in mind the cloud development model mentioned above, all UI capability (both front and backend) should be provided in a phased manner. [I would suggest initially that a topic listing page is the first deliverable](#topic-list-page), but before that can happen, [prerequisites](#prerequisites) need to be discussed, agreed and implemented.
+
+#### Prerequisites
+
+In order to provide any UI, the whole stack which will support it will need to be discussed and considered. This will include (but not be limited to):
+
+- Setting up the required UI build/packaging for deployment to Strimzi
```

Comment:
Currently the `strimzi-kafka-operators` repo has the build for both the operator code (essentially a maven build), the operator images and the packaging into zipped YAML, Helm charts etc. I suppose we might integrate this as follows:

I can see the `strimzi-ui` repo building the docker images, what we need to figure out is whether the installation bits also belong in `strimzi-ui`, or in `strimzi-kafka-operators`, and what the UX for installing the UI is. E.g. is it just a separate installation, or does it get installed with the CO, or is it installed by the CO. (I'm assuming you get a separate UI per CO, but maybe I'm wrong about that).

Reply:
> Currently the `strimzi-kafka-operators` repo has the build for both the operator code (essentially a maven build), the operator images and the packaging into zipped YAML, Helm charts etc. I suppose we might integrate this as follows:
> 
> I can see the `strimzi-ui` repo building the docker images, what we need to figure out is whether the installation bits also belong in `strimzi-ui`, or in `strimzi-kafka-operators`, and what the UX for installing the UI is. E.g. is it just a separate installation, or does it get installed with the CO, or is it installed by the CO. (I'm assuming you get a separate UI per CO, but maybe I'm wrong about that).

Absolutley - I imagine where the UI is deployed will be the key part of how the implementation will need to be done, but also the UX. As per a previous comment, I am open to what the community thinks is the best place to deploy a UI, and thus also the build/deploy implementation/UX for it as well.

Reply:
I would expect one UI per CO but maybe able to handling more Kafka clusters deployed by the same CO in different namespaces. Maybe the UI server should be configured to connect to different backend API; I am assuming that each Kafka cluster deployed in its own namespace will get its own backend server.

Reply:
> I would expect one UI per CO

Is that just an expectation or a suggested enforcement @ppatierno ? I could see a use-case for multiple UIs, one for each Kafka.

Reply:
@samuel-hawker What's the use-case you see? A kind of multitenancy? Can you elaborate a little bit more please?

Reply:
Yes I was thinking multi-tenancy. I dont have any sort of data saying this would actually be any specific users requirement. But from a practical engineering point of view if we design this to be installable multiple per cluster to begin with if the requirement ever came up it would be trivial to support as opposed to guaranteeing unique naming etc may be harder if we assume just one per cluster to begin with.  Just wondering what having omly one UI buys us as allowing for multiple configured as any team likes.